lib/chat_web/live/main_live/page/lobby.ex

```diff
@@ -91,24 +90,57 @@ defmodule ChatWeb.MainLive.Page.Lobby do
       PubSub.broadcast!(
         Chat.PubSub,
         @topic,
-        :room_request
+        {:room_request, room_hash, me |> Identity.pub_key() |> Utils.hash()}
       )
     end)
 
     socket
     |> assign_room_list()
   end
 
-  def approve_requests(socket) do
+  def approve_room_request(%{assigns: %{room_map: room_map}} = socket, room_hash, user_hash)
+      when is_map_key(room_map, room_hash) do
+    {_, _, {encrypted, blob}} =
+      room_hash
+      |> Rooms.approve_requests(room_map[room_hash])
+      |> Rooms.get_request(user_hash)
+
+    PubSub.broadcast!(
+      Chat.PubSub,
+      @topic,
+      {:room_request_approved, {encrypted, blob}, user_hash}
+    )
+
     socket
-    |> approve_joined_room_requests()
+  rescue
+    _ -> socket
   end
 
-  def join_rooms(socket) do
-    socket
-    |> join_approved_rooms()
+  def approve_room_request(socket, _, _), do: socket
+
+  def join_approved_room(
+        %{assigns: %{me: me, my_id: my_id, monotonic_offset: time_offset, room_map: room_map}} =
+          socket,
+        encrypted_room_identity,
+        user_hash
+      )
+      when my_id == user_hash do
+    new_room_identity = Rooms.decrypt_identity(encrypted_room_identity, me)
+
+    if new_room_identity in Map.values(room_map) do
```

Comment:
it is faster to `Map.has_key?(room_map, new_room_identity |> Identity.pub_key() |> Utils.hash())`